### PR TITLE
GH Actions Workflow for AWS Authentication 

### PR DESCRIPTION
**Pull request by nataliejschultz**
This is the second PR for e-mission/e-mission-docs#1008

The primary python script used in this work, email-config.py, is from  nataliejschultz:AWS-email-config . This PR aims to expand the functionality of email-config.py with automation. 

main.yml is the GitHub Actions workflow file that will run email-config.py. This workflow file consists of two jobs: 

- AssumeRoleAndCallIdentity, which accesses our IAM role and grants our workflow permissions
- build, which installs dependencies and runs email-config.py. 

The workflow should only run when a new config file (as a .json file) is added to the configs directory. The name of the new config file will be passed into email-config.py.

### CODE REVIEW

**Comment by nataliejschultz** [comment]
Current status: 

I struggled a lot today with passing the name of the newly modified config file to the main job in GitHub Actions (GHA), but got it to work. Have also spent a lot of time troubleshooting file path discrepancies between local and GHA runs, but I believe that's finally figured out. 

I've gotten the script to work in GHA all the way up to calling update_user_pool. I ran into yet another authentication issue, so I have to wait for our AWS liaison to get back to me tomorrow.

**Comment by nataliejschultz** [comment]
It worked!!!! 🎉🎉🎉

All of the AWS configuration has been figured out. I ran a push test (pushing an insignificant change to the Wyoming.nrel-op.json file) after the settings went through on Jianli's part, and I finally got my welcome email. I followed the link and logged in successfully.

I'm going to look over the code before putting it into ready for review.

File: README.md
Comment (nataliejschultz) [inline README.md:1]: I don't know why this file is in here. I didn't update it.

Comment (shankari) [reply README.md:1]: If you didn't update it, you can just revert the change using `git checkout`

File: configs/wyoming.nrel-op.json [hunk 1]
```diff
@@ -44,6 +44,7 @@
         "trip_end_notification": false
     },
     "admin_dashboard": {
+        "admin_access": "nschultz@nrel.gov",
```
Comment (shankari) [inline configs/wyoming.nrel-op.json:47]: Please revert this change since we don't actually want to give you access to the wyoming data

File: email_automation/welcome-template.txt
Comment (shankari) [inline email_automation/welcome-template.txt:1]: this file should be checked in as part of https://github.com/e-mission/nrel-openpath-deploy-configs/pull/45/files, right?

File: email_automation/email-config.py
Comment (shankari) [inline email_automation/email-config.py:1]: Ditto.

**Review by shankari — CHANGES_REQUESTED** [review]
Please clean up the PR so that it has only the changes that I need to review (aka only `main.yml`). And I would rename `main.yml` to have a more meaningful name.

File: email_automation/welcome-template.txt
Comment (nataliejschultz) [reply email_automation/welcome-template.txt:1]: There are identical versions of `welcome-template.txt` and `email-config.py` in both PRs #45 and #54. I had to add them here to test the`AWS-auth.yml`(nee`main.yml`) workflow since #45 hasn't been merged yet, and have made sure to keep the files the same while making changes along the way(see commit 9e5120472c51a1b4b69237d0223e9aa21d44407d) 

Since the intention was to have #45 merged first, I will remove the two duplicate files from this PR.

File: configs/wyoming.nrel-op.json (hunk 1 above)
Comment (nataliejschultz) [reply configs/wyoming.nrel-op.json]: Reverted :)

File: README.md
Comment (nataliejschultz) [reply README.md:1]: Fixed!! File no longer marked as changed in this PR!

File: email_automation/email-config.py
Comment (nataliejschultz) [reply email_automation/email-config.py:1]: Removed both files that should be merged from #45 first. The only file remaining is the workflow file. Runs [have succeeded ](https://github.com/e-mission/nrel-openpath-deploy-configs/actions/runs/7646138875)as-is, though it would be helpful to fill out the config file template and test it for real (especially if we get the user pool generation fully automated) . 

Since the workflow might change after the meeting tomorrow, I'm moving this back into `issues being worked on` for now. If there's no change after tomorrow, or if @shankari would prefer to close this PR and have me open another one for additional changes, I'll put it back into ready for review.

**Comment by nataliejschultz** [comment]
Was going through my merges and saw that this PR has not been merged. I'm not sure why.

File: .github/workflows/AWS-auth.yml [hunk 2]
```diff
@@ -0,0 +1,68 @@
+name: AWS Auth + Welcome Email
+
+on:
+  push:
+    branches:
+      - 'test_userpool_creation'
+    paths:
+        '**.json'
```
Comment (shankari) [inline .github/workflows/AWS-auth.yml:8]: Not a blocker, but I think this should really be `config/**.json` just in case we end up with json files elsewhere.

File: .github/workflows/AWS-auth.yml [hunk 3]
```diff
@@ -0,0 +1,68 @@
+name: AWS Auth + Welcome Email
+
+on:
+  push:
+    branches:
+      - 'test_userpool_creation'
```
Comment (shankari) [inline .github/workflows/AWS-auth.yml:6]: This needs to be changed. After the merge, we no longer want to run this only on the `test_userpool_creation` branch

File: .github/workflows/AWS-auth.yml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+      - name: List all changed files
+        id: config-file-name
+        run: |
+          for file in ${{ steps.get-changed-files.outputs.all_changed_files }}; do
+            if [[ "$file" == *nrel-op.json ]]; then
+              echo "The name of the config file is: ${file}."
+              echo "CONFIG_FILE=${file}" >> "$GITHUB_OUTPUT"
+            fi
+          done
```
Comment (shankari) [inline .github/workflows/AWS-auth.yml:43]: what happens if there were multiple changed files? This looks like it will return only the last value. Won't we then skip the modifications in the other changed files?

**Review by shankari — CHANGES_REQUESTED** [review]
Almost there, just a few high-level comments

**Comment by nataliejschultz** [comment]
Finishing up this project, which got pushed aside for the image project.

In adding functionality for addressing multiple changed files, I ran into an[ issue ](https://github.com/e-mission/nrel-openpath-deploy-configs/actions/runs/8745328251/job/24000121670)with the IAM role not allowing the usage of AdminDeleteUser with Cognito calls. 
```
botocore.exceptions.ClientError: An error occurred (AccessDeniedException) when calling the AdminDeleteUser operation: User: arn:aws:sts::***:assumed-role/***/GitHub_to_AWS_via_FederatedOIDC is not authorized to perform: cognito-idp:AdminDeleteUser on resource: arn:aws:cognito-idp:us-west-2:***:userpool/us-west-2_------ because no identity-based policy allows the cognito-idp:AdminDeleteUser action
```
Reached out to Jianli, who added the functionality to the role. 

Then, tried modifying two config files at once. nrel-commute had 0 users in the pool, and wyoming only had my other email present:

Screenshot 2024-06-11 at 3 50 36 PM

I modified the config files. For commute, I added my NREL email to the proper section. For Wyoming, I did the same, and did not list my other email that had an account already:
```

    "admin_access": [
      " nschultz@nrel.gov"
```

Pushing the changes ran the job:

Screenshot 2024-06-11 at 3 51 32 PM

As you can see, both of the changed config files were recognized, and the script ran for each one. 

My old email was removed from the Wyoming user pool, and the NREL email was added:
Screenshot 2024-06-11 at 3 50 47 PM

I got a welcome email, and was able to log in with the temporary password sent: 

Screenshot 2024-06-11 at 10 26 05 PM

Additionally, I was added to the commute pool and logged in with my temporary credentials:

Screenshot 2024-06-11 at 10 26 13 PM

**Comment by shankari** [comment]
This looks fine to me, but I have a couple of high-level comments:
1. Alas, we will not be able to create the accounts automatically when the environment is created (@Abby-Wheelis). This is because the config needs to be in place at the time that the environment is created, at least to test out the webapp, join page etc.
   - I wonder if there is some way to figure out that the environment has been created, and then launch this again.
   - Or maybe after the DNS changes, we no longer need to have the config merged first, need to check with the cloud services team
   - I do note that the script fails if the user pool has not been created, so this will work correctly after all
2. But this is still useful when the list of admin users is modified, which does happen fairly frequently, so I will go ahead and merge this.
3. We need to get AWS credentials set up for this, right? @nataliejschultz can you send out an email (internally) with the credentials required so that I can configure them?

**Review by shankari — APPROVED** [review]

**Comment by shankari** [comment]
@nataliejschultz please note that I am squash-merging these files to avoid commit churn.

**Comment by nataliejschultz** [comment]
> 3. We need to get AWS credentials set up for this, right? @nataliejschultz can you send out an email (internally) with the credentials required so that I can configure them?

This was merged from a branch that we created on the regular e-mission repo (rather than a fork), so I think the credentials should be configured already! We set it up this way so that I could test it properly with AWS.

**Comment by shankari** [comment]
I just changed the list of admin users 
https://github.com/e-mission/nrel-openpath-deploy-configs/commit/6922e3ccebdc2bfc056f082f9fe5689c4ff1b8d6

and the action ran successfully!

```
changed files string: configs/e-bikes-for-essentials.nrel-op.json
config file name configs/e-bikes-for-essentials.nrel-op.json
bmarkham@vailgov.com removed from pool.
...
bmarkham@vail.gov not in user pool! Creating account...
Account created! Sending welcome email.
```
